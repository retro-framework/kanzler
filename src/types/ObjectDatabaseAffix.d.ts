import { IStateErrorable, IAffix } from "./index";

/*
 * Redux State
 */
export interface IState extends IStateErrorable {
  readonly affix: IAffix;
}

/*
 * Component Props
 */
interface IPropsFromState extends IStateErrorable {
  affix: IAffix;
}

// tslint:disable-next-line:no-empty-interface
interface IPropsFromDispatch {
  // selectCheckpointFn: (affix: IAffix) => void;
}

// tslint:disable-next-line:no-empty-interface
export interface IProps extends IPropsFromState, IPropsFromDispatch {}
