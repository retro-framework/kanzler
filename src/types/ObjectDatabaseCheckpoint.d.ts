import { IStateErrorable, ICheckpoint } from './index'

/*
 * Redux State
 */
export interface IState extends IStateErrorable {
    readonly checkpoint: ICheckpoint
}

/*
 * Component Props
 */
interface IPropsFromState extends IStateErrorable {
    checkpoint: ICheckpoint
}

// tslint:disable-next-line:no-empty-interface
interface IPropsFromDispatch {
    selectCheckpointFn: (checkpoint: ICheckpoint) => void
}

// tslint:disable-next-line:no-empty-interface
export interface IProps extends IPropsFromState, IPropsFromDispatch {}