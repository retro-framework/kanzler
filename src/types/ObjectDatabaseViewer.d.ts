import { IStateErrorable, IStateLoadable, ICheckpoint } from './index'

/*
 * Redux State
 */
export interface IState extends IStateErrorable, IStateLoadable {
    readonly selectedHeadRefHash: string
    readonly checkpoints: ICheckpoint[]
} 

/*
 * Component Props
 */
interface IPropsFromState {
    checkpoints: ICheckpoint[]
    isLoading: boolean
    selectedHeadRefHash: string
}

// tslint:disable-next-line:no-empty-interface
interface IPropsFromDispatch {}

// tslint:disable-next-line:no-empty-interface
export type IProps = IPropsFromState | IPropsFromDispatch;