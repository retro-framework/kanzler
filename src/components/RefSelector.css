.RefSelector:before {
  content: "Head Ref:";
}

.RefSelector {
  padding: 5px 35px 5px 5px;
  /* font-size: 16px; */
  border: 1px solid #ccc;
  height: 34px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: url(http://www.stackoverflow.com/favicon.ico) 96% / 15% no-repeat #eee;
}

/* CAUTION: IE hackery ahead */
.RefSelector::-ms-expand {
  display: none;
  /* remove default arrow in IE 10 and 11 */
}

/* target Internet Explorer 9 to undo the custom arrow */x`
@media screen and (min-width:0\0) {
  .RefSelector {
    background: none\9;
    padding: 5px\9;
  }
}