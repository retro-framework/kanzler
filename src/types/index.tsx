export interface IStateLoadable {
    readonly loading: boolean
}

export interface IStateErrorable {
    readonly error?: Error | undefined
}

export interface IRefHash {
    readonly hash: string
    readonly name: string
}

export interface ICheckpoint {
    readonly hash: string
    readonly parentHashes: string[]
    readonly affixHash: string
    readonly commandDesc: string
    readonly summary: string
}