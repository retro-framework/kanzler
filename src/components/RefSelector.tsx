import * as React from 'react';

import { IProps } from '../types/RefSelector';

import './RefSelector.css';

class RefSelector extends React.Component<IProps> {
  public update = (e: any) => {
    this.props.handleChangedSelectedHeadRefHash(e.target.value);
  }
  public componentDidUpdate() {
    this.props.handleChangedSelectedHeadRefHash(this.props.selectedHash);
  }
  public render() {
    if (this.props.loading) {
      return "<span>Loading ...</span>"
    }
    return (
      <select className="RefSelector" value={this.props.selectedHash} onChange={this.update}>
        <option>Choose...</option>
        {this.options()}
      </select>
      );
  }
  private options():any {
    return (this.props && this.props.refs).map((ref: any) => <option key={ref.hash.concat(ref.name)} value={ref.hash}>{ref.name}</option>);
  }
}

export default RefSelector;
