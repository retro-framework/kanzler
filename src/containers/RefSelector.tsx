import { connect, Dispatch } from 'react-redux';

import * as actions from '../actions/';
import RefSelector from '../components/RefSelector';
import { IPropsFromDispatch, IPropsFromState } from '../types/RefSelector';
import IStoreState from '../types/store';

export function mapStateToProps(state: IStoreState): IPropsFromState {
    return {
        loading: state.refSelector.loading,
        refs: state.refSelector.refs,
        selectedHash: state.refSelector.selectedHash,
    };
}

async function getCheckpoints(startHash: string) {
    const getCheckpoint = async (hash: string) => {
        const result = await fetch(`/obj/${hash}`).then(response => response.json()).catch(e => console.error);
        if (result) {
            return Object.assign({ hash }, result);
        } else {
            return;
        }
    }
    const checkpoints: any[] = [];
    const firstCp = await getCheckpoint(startHash);
    if (!firstCp) {
        return checkpoints;
    }
    checkpoints.push(firstCp);

    let last = checkpoints[checkpoints.length - 1];
    while (last.parentHashes && last.parentHashes.length) {
        checkpoints.push(await getCheckpoint(checkpoints[checkpoints.length - 1].parentHashes[0]));
        last = checkpoints[checkpoints.length - 1];
    }
    return checkpoints;
}

function mapDispatchToProps(dispatch: Dispatch<actions.ISetSelectedHeadRefHash | actions.IODBVCheckpointsAvailable | actions.IODBVInvalidate>): IPropsFromDispatch {
    return {
        handleChangedSelectedHeadRefHash: (newSelectedHeadRefHash: string) => {
            dispatch(actions.odbvInvalidate());
            dispatch(actions.setSelectedHeadRefHash(newSelectedHeadRefHash));
            getCheckpoints(newSelectedHeadRefHash).then(checkpoints => {
                dispatch(actions.odbvCheckpointsAvailable(checkpoints));
            })
        }
    }
}

export default connect<IPropsFromState, IPropsFromDispatch, void>(mapStateToProps, mapDispatchToProps)(RefSelector);