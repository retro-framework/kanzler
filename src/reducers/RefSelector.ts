import * as actions from '../actions';
import { REF_LIST_ENTRIES_AVAILABLE, REF_LIST_INVALIDATE, SET_SELECTED_HEAD_REF_HASH } from '../constants/index';
import { IState } from '../types/RefSelector';

export default function (state: IState, action: actions.IRefListEntriesAvailable | actions.IInvalidateRefList | actions.ISetSelectedHeadRefHash): IState {
  switch (action.type) {
    case REF_LIST_ENTRIES_AVAILABLE:
      return { ...state, loading: false, refs: action.payload, selectedHash: action.payload[action.payload.length -1].hash };
    case REF_LIST_INVALIDATE:
      return { ...state, loading: true, refs: [] };
    case SET_SELECTED_HEAD_REF_HASH:
      return { ...state, selectedHash: action.payload };
  }
  return { ...state };
}