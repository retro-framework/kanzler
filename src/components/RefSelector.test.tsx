import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { IProps } from '../types/RefSelector';
import RefSelector from './RefSelector';

type TestCases = Array<{
  readonly name: string 
  readonly props: IProps
}>;

const cases: TestCases = [
  {
      name: 'displays nothing when no refs loaded', 
      props: {refs: [], selectedHash: "", loading: false, handleChangedSelectedHeadRefHash (_: string){  _ = _ } }
  },
  {
    name: 'displays loading span rather than dropdown when loading', 
    props: {refs: [], selectedHash: "", loading: true, handleChangedSelectedHeadRefHash (_: string){  _ = _ }}
  },
  {
    name: 'displays error nothing when no non-existent hash ref is defined', 
    props: {refs: [{hash: "abc123", name: "refs/heads/mainline"}], selectedHash: "blah blah", loading: false, handleChangedSelectedHeadRefHash (_: string){  _ = _ }}
  },
  {
    name: 'displays correctly when well defined hash ref is selected', 
    props: {refs: [{hash: "abc123", name: "refs/heads/mainline"}], selectedHash: "abc123", loading: false, handleChangedSelectedHeadRefHash (_: string){  _ = _ }}
  },
  {
    name: 'displays placeholder when no refs loaded and not loading', 
    props: {refs: [], selectedHash: "", loading: false, handleChangedSelectedHeadRefHash (_: string){  _ = _ }}
  }
];

cases.forEach(testCase => {

  test(testCase.name, () => {
    const component = renderer.create(
      <RefSelector {...testCase.props} />,
    );
      
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

  });
})