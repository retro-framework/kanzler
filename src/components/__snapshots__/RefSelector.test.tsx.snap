// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays correctly when well defined hash ref is selected 1`] = `
<select
  className="RefSelector"
  onChange={[Function]}
  value="abc123"
>
  <option>
    Choose...
  </option>
  <option
    value="abc123"
  >
    refs/heads/mainline
  </option>
</select>
`;

exports[`displays error nothing when no non-existent hash ref is defined 1`] = `
<select
  className="RefSelector"
  onChange={[Function]}
  value="blah blah"
>
  <option>
    Choose...
  </option>
  <option
    value="abc123"
  >
    refs/heads/mainline
  </option>
</select>
`;

exports[`displays loading span rather than dropdown when loading 1`] = `"<span>Loading ...</span>"`;

exports[`displays nothing when no refs loaded 1`] = `
<select
  className="RefSelector"
  onChange={[Function]}
  value=""
>
  <option>
    Choose...
  </option>
</select>
`;

exports[`displays placeholder when no refs loaded and not loading 1`] = `
<select
  className="RefSelector"
  onChange={[Function]}
  value=""
>
  <option>
    Choose...
  </option>
</select>
`;
